core/src/mast/mod.rs

```diff
@@ -191,6 +202,68 @@ impl MastForest {
         self[node_id].set_after_exit(decorator_ids)
     }
 
+    /// Merges all `forests` into a new [`MastForest`].
+    ///
+    /// Merging two forests means combining all their constituent parts, i.e. [`MastNode`]s,
+    /// [`Decorator`]s and roots. During this process, any duplicate or
+    /// unreachable nodes are removed. Additionally, [`MastNodeId`]s of nodes as well as
+    /// [`DecoratorId`]s of decorators may change and references to them are remapped to their new
+    /// location.
+    ///
+    /// For example, consider this representation of a forest's nodes with all of these nodes being
+    /// roots:
+    ///
+    /// ```text
+    /// [Block(foo), Block(bar)]
+    /// ```
+    ///
+    /// If we merge another forest into it:
+    ///
+    /// ```text
+    /// [Block(bar), Call(0)]
+    /// ```
+    ///
+    /// then we would expect this forest:
+    ///
+    /// ```text
+    /// [Block(foo), Block(bar), Call(1)]
+    /// ```
+    ///
+    /// - The `Call` to the `bar` block was remapped to its new index (now 1, previously 0).
+    /// - The `Block(bar)` was deduplicated any only exists once in the merged forest.
+    ///
+    /// The function also returns a vector of [`MastForestRootMap`]s, whose length equals the number
+    /// of passed `forests`. The indices in the vector correspond to the ones in `forests`. The map
+    /// of a given forest contains the new locations of its roots in the merged forest. To
+    /// illustrate, the above example would return a vector of two maps:
+    ///
+    /// ```text
+    /// vec![{0 -> 0, 1 -> 1}
+    ///      {0 -> 1, 1 -> 2}]
+    /// ```
+    ///
+    /// - The root locations of the original forest are unchanged.
+    /// - For the second forest, the `bar` block has moved from index 0 to index 1 in the merged
+    ///   forest, and the `Call` has moved from index 1 to 2.
+    ///
+    /// If any forest being merged contains an `External(qux)` node and another forest contains a
+    /// node whose digest is `qux`, then the external node will be replaced with the `qux` node,
+    /// which is effectively deduplication. Decorators are ignored when it comes to merging
+    /// External nodes. This means that an External node with decorators may be replaced by a node
+    /// without decorators or vice versa.
+    pub fn merge(
+        forests: impl IntoIterator<Item = MastForest>,
+    ) -> Result<(MastForest, Vec<MastForestRootMap>), MastForestError> {
```

Comment:
Question: do we actually need to take ownership of `MastForest`s? For example, could the iterator be over `<Item = &MastForest>`?

Also, thinking about the return type a bit more, maybe we should move the vector inside `MastForestRootMap`? Basically, we'd return `(MastForest, MastForestRootMap)`, but the `map_root()` would have the following signature:

```rust
pub fn map_root(&self, forest_idx: usize, root: &MastNodeId) -> Option<MastNodeId> {
    ...
}
```
The functionality should be the same but it looks a bit cleaner to me.